upf/upf_ipfix.c

```diff
@@ -0,0 +1,693 @@
+    {
+      kv->key[0] = 0;
+      kv->key[1] = 0;
+    }
+  else
+    {
+      kv->key[0] = collector_ip->ip.as_u64[0];
+      kv->key[1] = collector_ip->ip.as_u64[1];
+      if (collector_ip->version == AF_IP4)
+	kv->key[2] |= 1;
+    }
+}
+
+u32
+upf_ref_ipfix_context (bool is_ip4,
+		       upf_ipfix_policy_t policy,
+		       const ip_address_t * collector_ip)
+{
+  int rv;
+  vlib_thread_main_t *tm = &vlib_thread_main;
+  upf_ipfix_main_t *fm = &upf_ipfix_main;
+  clib_bihash_kv_24_8_t kv, value;
+  upf_ipfix_protocol_context_t *context;
+  /* Decide how many worker threads we have */
+  u32 num_threads = 1 /* main thread */  + tm->n_threads;
+
+  upf_ipfix_context_mk_key(is_ip4, policy, collector_ip, &kv);
+
+  if (PREDICT_TRUE
+      (!clib_bihash_search_24_8
+       (&fm->context_by_key, &kv, &value)))
+    {
+      context = pool_elt_at_index (fm->contexts, value.value);
+      context->ref_count++;
+      return value.value;
+    }
+
+  pool_get_zero (fm->contexts, context);
+
+  vec_validate (context->buffers_per_worker, num_threads - 1);
+  vec_validate (context->frames_per_worker, num_threads - 1);
+  vec_validate (context->next_record_offset_per_worker,
+		num_threads - 1);
+  kv.value = context - fm->contexts;
+  clib_bihash_add_del_24_8 (&fm->context_by_key, &kv, 1);
+
+  context->policy = policy;
+  context->is_ip4 = is_ip4;
+  /* lookup the exporter a bit later */
+  context->exporter_index = (u32) ~0;
+  context->ref_count = 1;
+  if (collector_ip)
+    ip_address_copy (&context->collector_ip, collector_ip);
+
+  rv = upf_ipfix_report_add_del (fm, UPF_IPFIX_OBSERVATION_DOMAIN_ID,
+				 kv.value, &context->template_id,
+				 is_ip4, true);
+  if (rv)
+      clib_warning ("couldn't add IPFIX report, perhaps "
+		    "the exporter has been deleted?");
+
+  return context - fm->contexts;
+}
+
+void
+upf_ref_ipfix_context_by_index (u32 cidx)
+{
+  upf_ipfix_main_t *fm = &upf_ipfix_main;
+  upf_ipfix_protocol_context_t *context;
+
+  context = pool_elt_at_index (fm->contexts, cidx);
+  context->ref_count++;
+}
+
+void
+upf_unref_ipfix_context_by_index (u32 cidx)
+{
+  int rv;
+  upf_ipfix_main_t *fm = &upf_ipfix_main;
+  clib_bihash_kv_24_8_t kv;
+  upf_ipfix_protocol_context_t *context;
+
+  context = pool_elt_at_index (fm->contexts, cidx);
+  if (--context->ref_count)
+    return;
+
+  upf_ipfix_context_mk_key(context->is_ip4, context->policy,
+			   &context->collector_ip,
+			   &kv);
+  clib_bihash_add_del_24_8 (&fm->context_by_key, &kv, 0 /* is_add */ );
+
+  rv = upf_ipfix_report_add_del (fm, UPF_IPFIX_OBSERVATION_DOMAIN_ID,
+				 cidx, &context->template_id,
+				 context->is_ip4, false);
+  if (rv)
+    clib_warning ("couldn't remove IPFIX report, perhaps "
+		  "the exporter has been deleted?");
+
+  vec_free (context->buffers_per_worker);
+  vec_free (context->frames_per_worker);
+  vec_free (context->next_record_offset_per_worker);
+  pool_put (fm->contexts, context);
```

Comment:
pool_put is not thread safe. The unref method can be called from multiple threads, when this happens, the put will corrupt the pool.

Reply:
Added spinlock